### ISSUE TITLE
Memory-Leak in prometheus-operator

### ISSUE THREAD

**Issue by steigr** [coreos/prometheus-operator#483]
**What did you do?**

Fresh k8s/tectonic 1.6.7-tectonic.1 install with tectonic-installer on 4 nodes ( 3 masters + 1 worker).

**What did you expect to see?**

prometheus-operator should run continuously.

**What did you see instead? Under which circumstances?**

the operator memory usage grows and get eventually kill because of OOM. k8s reschedules the pod.
<img width="531" alt="memory-leak" src="https://user-images.githubusercontent.com/186793/28345213-12810d5e-6c29-11e7-83d9-646e458c1b0b.png">

**Environment**

k8s/tectonic 1.6.7-tectonic.1, 3 masters with 2 threads and 2gb ram, 1 worker with 8 threads and 48 gb ram. The nodes are connected over a VPN (wireguard). Latency between masters is low (~1ms). The worker is reachable within 15-20ms.

* Kubernetes version information:

```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.4", GitCommit:"d6f433224538d4f9ca2f7ae19b252e6fcb66a3ae", GitTreeState:"clean", BuildDate:"2017-05-19T20:41:24Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.7+coreos.0", GitCommit:"c8c505ee26ac3ab4d1dff506c46bc5538bc66733", GitTreeState:"clean", BuildDate:"2017-07-06T17:38:33Z", GoVersion:"go1.7.6", Compiler:"gc", Platform:"linux/amd64"}
```

* Kubernetes cluster kind: 

via tectonic-installer 1.6.7

* Manifests:

[tectonic-installer/.../prometheus-k8s.yaml](https://github.com/coreos/tectonic-installer/blob/master/modules/tectonic/resources/manifests/monitoring/prometheus-k8s.yaml)

* Prometheus Operator Logs:

```
...
ts=2017-07-19T00:26:55Z caller=operator.go:972 component=prometheusoperator msg="updating config skipped, no configuration change"
ts=2017-07-19T00:26:55Z caller=operator.go:584 component=prometheusoperator msg="update handler" old=324451 cur=324458
ts=2017-07-19T00:26:55Z caller=operator.go:321 component=alertmanageroperator msg="update handler" old=324451 cur=324458
ts=2017-07-19T00:26:55Z caller=operator.go:619 component=prometheusoperator msg="sync prometheus" key=tectonic-system/k8s
ts=2017-07-19T00:26:55Z caller=operator.go:972 component=prometheusoperator msg="updating config skipped, no configuration change"
ts=2017-07-19T00:27:01Z caller=operator.go:619 component=prometheusoperator msg="sync prometheus" key=tectonic-system/k8s
ts=2017-07-19T00:27:01Z caller=operator.go:972 component=prometheusoperator msg="updating config skipped, no configuration change"
ts=2017-07-19T00:27:02Z caller=operator.go:584 component=prometheusoperator msg="update handler" old=324458 cur=324544
ts=2017-07-19T00:27:02Z caller=operator.go:321 component=alertmanageroperator msg="update handler" old=324458 cur=324544
ts=2017-07-19T00:27:02Z caller=operator.go:619 component=prometheusoperator msg="sync prometheus" key=tectonic-system/k8s
ts=2017-07-19T00:27:02Z caller=operator.go:972 component=prometheusoperator msg="updating config skipped, no configuration change"
...
```

**Comment by brancz** [comment]
Thanks a lot for reporting! I don't seem to have the same behavior on my cluster, but I will definitely investigate!

**Comment by colhom** [comment]
I am also observing steadily growing memory usage for this pod as well, though the rate of growth is much slower than what @steigr is observing. It is consistent and roughly linear w.r.t to time though

Approx. 24 hour period on an essentially idle cluster, seeing memory usage grow from ~20M to ~150M

![prom-operator-mem-graph](https://user-images.githubusercontent.com/1075028/28469988-fd1920d0-6dec-11e7-81a2-b8cde38ca1d9.png)

Will keep an eye on this and see what happens when it hits the limit.

**Comment by colhom** [comment]
@steigr what platform are you running tectonic on?

**Comment by brancz** [comment]
I remember we've seen memory leaks through client-go before, but everything is speculation until we measure. I'll make sure to add profiling endpoints next week (it's already EOD here in Germany :) ) and then we can investigate more clearly.

**Comment by brancz** [comment]
With #504 merged we can now start profiling. An image from current master is available under `quay.io/coreos/prometheus-operator:master`. Could you run this image and take a heap profile once you see high memory usage again?

If you're running Tectonic and have the experimental flag on (aka self updating Prometheus), you need to first disable that the Tectonic Prometheus Operator ensures that all manifests are overridden with what it thinks is the correct state. You can do that by setting the `paused` field in the `tectonic-monitoring` AppVersion to `true`. I'd suggest doing that with:

```
$ kubectl -n tectonic-system edit appversion tectonic-monitoring
```

Then update change the version of the Prometheus Operator to `master`:

```
$ kubectl -n tectonic-system edit deploy prometheus-operator
```

Once saved and deployed, figure out what you Prometheus Operator pod is:

```
$ kubectl -n tectonic-system get pods -lk8s-app=prometheus-operator       
NAME                                   READY     STATUS    RESTARTS   AGE
prometheus-operator-2489471810-13628   1/1       Running   0          48s
```

Then start `kubectl proxy` and then run the `pprof tool` and write the heap profile in svg format to disk with `svg > heap-inuse.svg`:

```
$ go tool pprof -inuse_space http://localhost:8001/api/v1/proxy/namespaces/tectonic-system/pods/prometheus-operator-2489471810-13628:8080/debug/pprof/heap                                                          
Fetching profile from http://localhost:8001/api/v1/proxy/namespaces/tectonic-system/pods/prometheus-operator-2489471810-13628:8080/debug/pprof/heap
Saved profile in /home/brancz/pprof/pprof.operator.localhost:8001.alloc_objects.alloc_space.inuse_objects.inuse_space.002.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) svg > heap-inuse.svg
Generating report in heap-inuse.svg
(pprof) quit
```

Then quit the profiling tool and upload the svg here :slightly_smiling_face: .

Any profiling attempt highly appreciated! I just started a new cluster and will see if I have a similar growth and will try profiling myself as well then.

**Comment by brancz** [comment]
So my cluster has been running for about a day and I can't seem to be able to reproduce this.

While the memory usage reported by cAdvisor seems to be growing:

![screenshot from 2017-07-25 11-11-26](https://user-images.githubusercontent.com/4546722/28564905-d2146304-712a-11e7-98f7-fea40fdbfc44.png)

The memory usage reported by the go runtime seem to be relatively stable:

![screenshot from 2017-07-25 11-16-31](https://user-images.githubusercontent.com/4546722/28564929-e67d5dc8-712a-11e7-87c7-58977165d000.png)

And neither heap in use bytes or idle bytes suggest that this is something happening in the Prometheus Operator:

![screenshot from 2017-07-25 11-22-24](https://user-images.githubusercontent.com/4546722/28565125-909bb336-712b-11e7-8e28-af65f8e05c4f.png)
![screenshot from 2017-07-25 11-22-08](https://user-images.githubusercontent.com/4546722/28565126-90c9a99e-712b-11e7-8d36-a9be3000fa6a.png)

However, my cluster might not be representative as I'm not running any actual workload on it, so I'm relying on you two reporting back. Thanks!

**Comment by alepaez** [comment]
Hello guys.

I'm experiencing the same issue. Cluster version 1.6.7-tectonic1
Got worst when I got my own apps running on tectonic.

![prom-operator](https://user-images.githubusercontent.com/1220168/28577204-7f101ff2-712c-11e7-8313-32b566749a4a.png)

Now I'm getting blank spaces on my metrics

![prom-metrics-failing](https://user-images.githubusercontent.com/1220168/28577327-d77abc9c-712c-11e7-90b0-6803826eb2f3.png)

Thank you!

**Comment by brancz** [comment]
> Got worst when I got my own apps running on tectonic.

That's what I suspected. Unfortunately it's hard for me to reproduce that.

@alepaez could you follow the above steps to try out latest master of the Prometheus Operator? That way we can take profile it and find out what's actually happening.

Thanks a lot for reporting!

**Comment by alepaez** [comment]
prometheus-operator updated successfully, but got an error taking heap in use profile:

```bash
kubectl -n tectonic-system get pods -lk8s-app=prometheus-operator
NAME                                  READY     STATUS    RESTARTS   AGE
prometheus-operator-948269252-1vtwh   1/1       Running   2          19m
```
```bash
go tool pprof -inuse_space http://localhost:8001/api/v1/proxy/namespaces/tectonic-system/pods/prometheus-operator-948269252-1vtwh:8080/debug/pprof/heap
Fetching profile from http://localhost:8001/api/v1/proxy/namespaces/tectonic-system/pods/prometheus-operator-948269252-1vtwh:8080/debug/pprof/heap
parsing profile: unrecognized profile format
```

**Comment by brancz** [comment]
@alepaez did you run `kubectl proxy`? Make sure it's still running (maybe through `kubectl proxy &`) so that the request is actually proxied.

**Comment by alepaez** [comment]
Yea, kubectl proxy was running. Another try with kubectl proxy &
```bash
kubectl proxy &
[1] 12021
Starting to serve on 127.0.0.1:8001
kubectl -n tectonic-system get pods -lk8s-app=prometheus-operator
NAME                                  READY     STATUS    RESTARTS   AGE
prometheus-operator-948269252-1vtwh   1/1       Running   14         2h
go tool pprof -inuse_space http://localhost:8001/api/v1/proxy/namespaces/tectonic-system/pods/prometheus-operator-948269252-1vtwh:8080/debug/pprof/heap
Fetching profile from http://localhost:8001/api/v1/proxy/namespaces/tectonic-system/pods/prometheus-operator-948269252-1vtwh:8080/debug/pprof/heap
parsing profile: unrecognized profile format
```
Get connection refused if run pprof without kubectl proxy

**Comment by brancz** [comment]
Which version of go are you using? I'm using 1.8 and that's also what the Prometheus Operator was compiled with. They might need to be in sync. Could you check that?

**Comment by alepaez** [comment]
Updated to go 1.8 and worked like a charm! :D 

[heap-inuse.zip](https://github.com/coreos/prometheus-operator/files/1174576/heap-inuse.zip)

**Comment by brancz** [comment]
Thanks! That suggests that only ~12Mb are in use. Could you take another profile with `-alloc_space` instead of `-inuse_space`?

**Comment by alepaez** [comment]
Sure! :)
[heap-alloc.zip](https://github.com/coreos/prometheus-operator/files/1176892/heap-alloc.zip)

**Comment by brancz** [comment]
Thanks! Interesting, it seems to all be in serializing/deserializing objects. We're in the process of upgrading to the latest versions of the Kubernetes libraries, that might already do the trick. If not we need to think of strategies to limit the objects we are list/watching to fill our cache, this is assuming that the sheer amount of objects in the cache are causing this amount of memory usage.

**Comment by alepaez** [comment]
Nice @brancz, upgrading prometheus-operator when a new version is release will do the trick?

**Comment by brancz** [comment]
Hopefully yes. The client-go version is updated, so that might already do the trick, but if not we have other options to explore if needed.

**Comment by alepaez** [comment]
@brancz This update is already available?

**Comment by brancz** [comment]
We're working hard towards a `v0.12.0` release. We can't promise a date though right now. It shouldn't be far away though.

**Comment by alepaez** [comment]
Hey @brancz 
Any news on this issue? Really missing my kubernetes metrics :(

**Comment by brancz** [comment]
We actually just released v0.12.0, I'm hoping updating client go just fixed the issue, if not I'll start profiling more heavily, as this badly needs to be fixed.

**Comment by alepaez** [comment]
Updated to v0.12.0 but It is not starting

```
ts=2017-08-25T11:59:30Z caller=operator.go:148 component=alertmanageroperator msg="connection established" cluster-version=v1.6.7+coreos.0
ts=2017-08-25T11:59:30Z caller=operator.go:231 component=prometheusoperator msg="connection established" cluster-version=v1.6.7+coreos.0
ts=2017-08-25T11:59:30Z caller=main.go:148 msg="Unhandled error received. Exiting..." err="creating CRDs failed: creating CRD: Prometheus: User \"system:serviceaccount:tectonic-system:prometheus-operator\" cannot create customresourcedefinitions.apiextensions.k8s.io at the cluster scope. (post customresourcedefinitions.apiextensions.k8s.io)"
```

![error-prometheus-operator](https://user-images.githubusercontent.com/1220168/29713197-e593d67e-8973-11e7-9619-ee061b980e65.png)

**Comment by brancz** [comment]
Excerpt from the release notes:

> NOTE: The RBAC roles of the Prometheus Operator need to be adapted to allow the automatic migration to succeed. 

:slightly_smiling_face: just apply the RBAC roles from this repo and you should be good to go.

**Comment by brancz** [comment]
We've been able to identify the memory leak, and fixed it with #629 and #630.

We've already released `v0.11.2` to fix this for `v1.6.x` Kubernetes clusters. A new version for `v1.7.x` clusters is coming up soon.

Closing this issue here.
